core/src/main/java/hudson/model/Queue.java

```diff
@@ -1427,6 +1438,20 @@ public void maintain() {
                 }
             }
 
+            //this is to solve OSS-192. We iterate on the list of flyweight tasks and execute them
+            List<FlyWeightItem> flyweightItems = new ArrayList<>(flyWeightTasks.values());
+            for (FlyWeightItem f : flyweightItems) {
+                if (!isBuildBlocked(f)) {
+                    // ready to be executed
+                    Runnable r = makeBuildable(new BuildableItem(f));
+                    if (r != null) {
+                        f.leave(this);
+                        r.run();
+                        updateSnapshot();
```

Comment:
I think it would be good to add a comment here explaining what's happening; how state is changing etc.